### PR TITLE
`ultralytics 8.0.196` instance-mean Segment loss

### PR DESCRIPTION
**Pull request by glenn-jocher**
Contributed by @yermandy 

### Summary
🔥🎨♻️

This pull request improves the code quality and readability of the `v8SegmentationLoss` class in `ultralytics/utils/loss.py`. It eliminates an unnecessary attribute, creates a new method for a common operation, and modifies the existing methods to use the new method.

> _`v8SegmentationLoss`_
> _Simpler and more reusable_
> _A fresh spring cleaning_

### Walkthrough
*  Remove unused attribute `self.nm` from `v8SegmentationLoss` class ([link](https://github.com/ultralytics/ultralytics/pull/5285/files?diff=unified&w=0#diff-d2ac3773214d5694c3286cf9534fc4914555f1a3f5c5ae4a99c6aff00c75892dL215))
*  Refactor segmentation loss calculation for a batch of images into a separate method `calculate_segmentation_loss` in `loss.py` ([link](https://github.com/ultralytics/ultralytics/pull/5285/files?diff=unified&w=0#diff-d2ac3773214d5694c3286cf9534fc4914555f1a3f5c5ae4a99c6aff00c75892dL271-R272))
*  Modify `single_mask_loss` method and add `calculate_segmentation_loss` method to `v8SegmentationLoss` class in `loss.py` ([link](https://github.com/ultralytics/ultralytics/pull/5285/files?diff=unified&w=0#diff-d2ac3773214d5694c3286cf9534fc4914555f1a3f5c5ae4a99c6aff00c75892dL298-R361))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements and code refactoring for YOLO's segmentation loss calculations.

### 📊 Key Changes
- Single mask loss calculation and segmentation loss now account for batch size.
- Refactored `single_mask_loss` method and introduced `calculate_segmentation_loss` with concise and improved documentation.
- Removed an unnecessary attribute initialization from the segmentation loss class.
- Improved in-line commenting and updated the code to match Python formatting standards.

### 🎯 Purpose & Impact
- The contribution offers a cleaner and better-documented code, which enhances maintainability and readability.
- The improved calculation methods in the segmentation loss may lead to more accurate training and, indirectly, better model performance.
- Streamlined code could provide a slightly faster execution during training, especially for single mask loss calculation thanks to tensor operations.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
Started comparison training with PR vs main in https://wandb.ai/glenn-jocher/Segment

**Comment by yermandy** [comment]
@glenn-jocher hello, thank you for your verification. As I mentioned, it should give 100% identical results but I see the strange thing that it does not run faster for you. Are you sure that you are running the correct code?

**Comment by yermandy** [comment]
@glenn-jocher Also, as we saw previously, averaging over objects in batch might improve performance compared to the current approach of averaging over objects within an image. I also wanted to verify it for the segmentation loss, here is the comparison, please, take a look:

https://wandb.ai/yermandy/seg-yolov8

**Comment by yermandy** [comment]
> @glenn-jocher hello, thank you for your verification. As I mentioned, it should give 100% identical results but I see the strange thing that it does not run faster for you. Are you sure that you are running the correct code?

Or maybe it is running not long enough yet

**Comment by yermandy** [comment]
> @glenn-jocher Also, as we saw previously, averaging over objects in batch might improve performance compared to the current approach of averaging over objects within an image. I also wanted to verify it for the segmentation loss, here is the comparison, please, take a look:
> 
> https://wandb.ai/yermandy/seg-yolov8

Unfortunately, you merged my branch and I cannot push to ultralytics origin. But it can be done by changing these three lines:

Screenshot 2023-10-09 at 7 36 00

**Comment by glenn-jocher** [comment]
Thanks for your suggestions, @yermandy. Indeed, averaging over objects in a batch could potentially improve performance. We'll take your ideas into consideration and evaluate their effectiveness. Regarding your query about the division by 0, you are correct. Typically, if the ground truth mask has no objects (i.e., `fg_mask.sum() == 0`), then we would expect the loss to be 0 as well. We appreciate your input, and we'll maintain these points as we iterate and improve the model's performance. Thank you!

**Comment by glenn-jocher** [comment]
@yermandy wandb shoes faster speeds. I see 10:58 per epoch for PR vs 11:24 per epoch for main branch with 3x GPUs on DDP.

**Comment by glenn-jocher** [comment]
> https://wandb.ai/yermandy/seg-yolov8

Ah perfect! Ok since the runs are indeed identical I can cancel the current PR branch and start training after a new commit with your updates. I'll do that today!

**Comment by glenn-jocher** [comment]
@yermandy changes applied in https://github.com/ultralytics/ultralytics/pull/5285/commits/2461635d49be049fa76dd22e6e0f5a3239730a24 and new training started in https://wandb.ai/glenn-jocher/Segment

**Comment by glenn-jocher** [comment]
@yermandy based on initial positive results of experiment I've gone ahead and merged this to main!

Thank you for your contributions!! 

I'll add retraining our Segment models to our TODOs here so we can provide new models retrained with this loss in the next few weeks.

@Laughing-q @AyushExel

**Comment by yermandy** [comment]
@glenn-jocher very good, nice to hear!

**Comment by yermandy** [comment]
Refactoring and speeding up computations for all losses was a crucial step for training multi-task model more efficiently

**Comment by glenn-jocher** [comment]
@yermandy thank you for your suggestion. We understand that efficient computation plays a crucial role in training multi-task models. Optimizing the computation for all loss functions, through refactoring and other means, can indeed speed up the process and make the training more efficient. We always strive for efficiency in our models, including YOLOv8, and we appreciate your insights in this matter. We will certainly take your suggestions into account in our future versions and updates. Thank you for your contribution to the Ultralytics community!
